spec/compiler/semantic/hooks_spec.cr

```diff
@@ -36,6 +36,27 @@ describe "Semantic: hooks" do
       ") { int32 }
   end
 
+  it "evaluate included macro on included module context" do
+    assert_type("
+      module Foo
+        A = 1
+
+        macro included
+          def self.foo
+            {{A}}
+          end
+        end
+      end
+
+      class Bar
+        A = 1.5
+        include Foo
+      end
+
+      Bar.foo
+      ") { int32 }
```

Comment:
I think should be float64. The type should be looked up in the including type. That's why we should use `current_type`.

Note that the hook is already executed in the **including** type, for example if you add a method to it. So type lookup should also be done from the including type.

Reply:
However, the hook name is just `included`. So, it seems intuitive that `included` module executes it.

In fact, Ruby's `included` hook works in such way:

```ruby
module Foo
  A = :Foo

  def self.included(klass)
    p A
  end
end

class Bar
  A = :Bar
  include Foo
end

# output:
# :Foo
```

Reply:
Yes, but in Ruby you get the `klass` as the including type, we don't do that in Crystal because we don't have runtime `define_method` and other stuff.

As I said, the current semantic should be revised before we try to change it or fix/brake things.

Reply:
See #5354. I consider `included` hook is a feature that it appends `Foo.included` automatically at `include Foo` if `Foo` has `included` macro, so this fix is correct in this view.

Reply:
I'm not sure #5354 is correct. It's unintuitive that the macro looks up type in the call scope and not the scope where the macro is defined.

All of this to say: we should formalize the semantics that we want here and then implement it. Right now this was all done with guesses and heuristics. This PR might fix something for some, I'm sure it will break code for others and we can't do that right before 1.0.0, specially not without understanding what we really want.